implementations/rust/ockam/ockam_command/src/base_command.rs

```diff
@@ -524,86 +524,98 @@ impl Display for ReadStatus {
+    fn validate(
+        &self,
+        ctx: &mut ValidationContext,
+    ) -> std::result::Result<ValidationResult, ReadlineError> {
+        let input = ctx.input();
+        if input.starts_with("\"\"\"") {
+            if input.ends_with("\n\"\"\"") {
+                return Ok(ValidationResult::Valid(None));
+            } else {
+                return Ok(ValidationResult::Incomplete);
+            }
+        }
+        Ok(ValidationResult::Valid(None))
+    }
 
-            loop {
-                read_buffer.clear();
-                let mut cancel_rx = cancel_rx.resubscribe();
-                let read_result = tokio::select! {
-                    result = Self::read_input(&mut stdin, &mut read_buffer) => result,
-                    _ = cancel_rx.recv() => {
-                        read_buffer = ":q\n".to_string();
-                        Ok(1)
-                    },
-                };
+    fn validate_while_typing(&self) -> bool {
+        false
+    }
+}
+impl Helper for MultiLineValidatorHelper {}
+
+impl RustylineHandler {
+    fn start() -> RustylineHandle {
+        let (next_s, next_r) = tokio::sync::mpsc::channel(1);
+        let (lines_s, lines_r) = tokio::sync::mpsc::channel(1);
+
+        // We run rustyline on their own blocking thread. The thread wait for an input line to be requested,
+        // then read it, and deliver it. While waiting for agent response there is no line requested, so readline
+        // is not active. If the thread thread terminate, the entire REPL exit.
+        let _join_handle = tokio::task::spawn_blocking(move || {
+            let _ = RustylineHandler::run_repl(next_r, lines_s);
+        });
+        RustylineHandle {
+            lines: lines_r,
+            next: next_s,
+        }
+    }
 
-                match read_result {
-                    Ok(0) => break, // EOF
-                    Ok(_n) => {
-                        if read_buffer.trim().is_empty() {
-                            continue;
-                        }
-                        let _ = tx.send(read_buffer.clone());
-                        if is_quit_sequence(&read_buffer) {
-                            break;
-                        }
-                    }
-                    Err(e) => {
-                        eprintln!("Error reading from stdin: {}", e);
+    fn run_repl(mut next_r: Receiver<()>, lines_s: Sender<String>) -> std::result::Result<(), ()> {
+        fn rustyline_error(err: ReadlineError) {
+            eprintln!("Repl error  {:?}", err);
+        }
+        let mut rl = Editor::new().map_err(rustyline_error)?;
+        rl.set_max_history_size(50).map_err(rustyline_error)?;
+        let helper = MultiLineValidatorHelper {};
+        rl.set_helper(Some(helper));
+        rl.load_history("history.txt").map_err(rustyline_error)?;
+        loop {
+            // If we can't read from the channel, exit
+            next_r.blocking_recv().unwrap();
+            let readline = rl.readline("> ");
```

Comment:
We don't receive the initial " >" from the server anymore? Or why do we need this?

Reply:
the input line is taken over by the library.  We do receive that from server still (we could remove if not using nc/telnet),  but what is displayed is whatever the prompt is here.